staging/src/k8s.io/apiserver/pkg/registry/rest/update.go

```diff
@@ -83,35 +83,43 @@ func validateCommonFields(obj, old runtime.Object, strategy RESTUpdateStrategy)
 // BeforeUpdate ensures that common operations for all resources are performed on update. It only returns
 // errors that can be converted to api.Status. It will invoke update validation with the provided existing
 // and updated objects.
+// It sets zero values only if the old object had a zero value for the respective field.
 func BeforeUpdate(strategy RESTUpdateStrategy, ctx context.Context, obj, old runtime.Object) error {
 	objectMeta, kind, kerr := objectMetaAndKind(strategy, obj)
 	if kerr != nil {
 		return kerr
 	}
+
+	oldMeta, err := meta.Accessor(old)
+	if err != nil {
+		return err
+	}
+
 	if strategy.NamespaceScoped() {
 		if !ValidNamespace(ctx, objectMeta) {
 			return errors.NewBadRequest("the namespace of the provided object does not match the namespace sent on the request")
 		}
-	} else {
+	} else if len(oldMeta.GetNamespace()) == 0 {
 		objectMeta.SetNamespace(metav1.NamespaceNone)
 	}
+
 	// Ensure requests cannot update generation
-	oldMeta, err := meta.Accessor(old)
-	if err != nil {
-		return err
-	}
 	objectMeta.SetGeneration(oldMeta.GetGeneration())
 
 	// Ensure Initializers are not set unless the feature is enabled
 	if !utilfeature.DefaultFeatureGate.Enabled(features.Initializers) {
+		if oldMeta.GetInitializers() == nil {
```

Comment:
unstructured accessor actually does the right thing with initializers already, so can drop this change